### ISSUE TITLE
Groups are randomly sorted when added via add_to_group()

### ISSUE THREAD

**Issue by stryker313** [godotengine/godot#51932]
### Godot version

 v3.3.3.stable.official

### System information

Windows 10 GTX 1060 6gb Intel 4690k

### Issue description

If you add a node to multiple groups, then call get_groups(), the order of the groups in the array is not constant, or consistent with how you called add_to_group. This can be bad if you are designing something with the order of the groups in mind

### Steps to reproduce

Create a node
Add the node to 4 groups in the _ready
in the physics or _ready, print(get_groups())
Groups will never be in the same order.

### Minimal reproduction project

[Godot Groups Error.zip](https://github.com/godotengine/godot/files/7024192/Godot.Groups.Error.zip)

Simply just run the file, and check the console. The order of the groups should correspond to the order in which they were added, but they are randomized everytime you run the game

**Comment by Calinou** [comment]
One way to resolve this would be to sort the groups array every time it's modified. However, doing so can have a noticeable performance cost.

**Comment by stryker313** [comment]
Yeah. Its just weird that the get_groups() would randomize like that. Ideally if you append(a),append(b),append(c) to an array, it should come out as [a,b,c], so unless there is a particular reason that get_groups() randomizes its kinda funky that it would do that. Is there a reason it functions like that, or was it just an oversight?

**Comment by kleonc** [comment]
Groups the given `Node` belongs to are stored in a `Map` (which is an unordered data structure):
https://github.com/godotengine/godot/blob/16642e28380cf5513fad6ea8591f809e42820382/scene/main/node.h#L112

So every insertion/removal can potentially change the order in that `Map` and thus the result of `get_groups` method might change:
https://github.com/godotengine/godot/blob/16642e28380cf5513fad6ea8591f809e42820382/scene/main/node.cpp#L2740
https://github.com/godotengine/godot/blob/16642e28380cf5513fad6ea8591f809e42820382/scene/main/node.cpp#L1647-L1665


Simply changing `Map` to `OrderedHashMap` (the one `Dictionary` uses exactly for the same reason) would make the order of insertion be preserved. It would probably be a little slower but not sure if that would by anyhow impactable (probably not).

**Comment by rxlecky** [comment]
On one hand, I understand where the confusion is coming from, on the other hand, order of groups which a node belongs to shouldn't matter.

> This can be bad if you are designing something with the order of the groups in mind

This is a **very** hypothetical scenario - I even dare to say, if you design a system that depends on the order of the groups it's a bad design to begin with. Do you have a real use-case where the order of groups matters and where there isn't a better solution? If not, this argument lives in a what-if land and it doesn't warrant changing the engine code - especially if there's a performance reason for why it's made the way it is.

**Comment by mhilbrunner** [comment]
Note: With https://github.com/godotengine/godot/pull/52310, this is now documented behaviour. We may want to close this issue if we don't have any good reason for changing this (and thus eating a possible performance penalty).
